### ISSUE TITLE
Adding a Feature to TERRAIN_OCEAN plot allows embarked unit to cross ocean before astronomy

### ISSUE THREAD

**Issue by Tekamthi** [LoneGazebo/Community-Patch-DLL#9973]
_1. Mod Version (X.Y.Z). Current Version: 3.7.5_

3.7.5

_2. DirectX Version_

_3. List of Other Mods_

MB+ v37

_4. Describe the Issue_

The mod above uses lua to add a storm feature (it adds an invisible feature and visible improvement simultaneously, but the feature is the problem) to ocean and coast plots.  Pre-astronomy embarked units are able to enter the feature plot in ocean (featureless ocean still correctly blocked).  Generalizing, a feature in an ocean plot should not allow pre-astronomy embarked movement.

______________________________________________________________________________________
_5. Save Game From 1 Turn Before (ALWAYS ATTACH THIS IF POSSIBLE)_

_6. Logs (ALWAYS ATTACH THESE IF POSSIBLE)_

_7. CvMiniDump.dmp File (ATTACH IF REPORTING A GAME CRASH)_

_8. Steps to reproduce the Issue (Optional)_

_9. Screenshots of the Issue (Optional)_

**Comment by Tekamthi** [comment]
Screenshot reference illustrating this issue -- the "storm" plot feature makes ocean unintentionally passable:
![Screenshot (369)](https://github.com/LoneGazebo/Community-Patch-DLL/assets/29668026/0922127c-b5a8-41ea-9d5c-22e61190acc6)
![Screenshot (368)](https://github.com/LoneGazebo/Community-Patch-DLL/assets/29668026/0ebf35c8-2c04-468a-9d55-08ac7011641e)

**Comment by Tekamthi** [comment]
I could write a .lua restriction for this myself but the necessary fields in UnitPromotions_Features appear to be borked.. opened a [separate issue](https://github.com/LoneGazebo/Community-Patch-DLL/issues/9972) for that

would be nice if terrain_ocean was always treated as terrain_ocean, feature or not, for pre-astro movement rules.  Granted this currently only relevant to modding, but i imagine this will be expected behavior to anyone else that tries to put features in ocean in any other mods as well, if that ever happens

**Comment by ilteroi** [comment]
should be fixed ... but i can't test

**Comment by Tekamthi** [comment]
> should be fixed ... but i can't test

I am trying in 3.7.8, but issue persists... is the tag intended as the next next release?

**Comment by ilteroi** [comment]
ah yes, bad timing. didn't know a release was in fact imminent

**Comment by ilteroi** [comment]
finally go around to test it ... fix didn't work.

in fact, i don't understand how it's possible that embarked and naval units behave differently here. the relevant line is (if i'm correct ...)

```
if(enterPlot.getFeatureType() != NO_FEATURE && isFeatureImpassable(enterPlot.getFeatureType()))
{
	return false;
}
```

so there is no difference ...

**Comment by Tekamthi** [comment]
yeah i tested on 3.7.9 just now -- still persists.  At the modding level, the naval unit's 'cannot enter ocean' promotion blocks access to these plots, whereas the 'embarkation' promotion controls the embarked unit's access -- the latter does some kind of tech check somewhere and is not controlled via the promotionfeatures/terrains tables. I can only guess what this means in .dll

I believe I can workaround this in my mod for now with some kind of promo to control access the same way as naval units, ie will just loop all land units and add this promo, then take it off after astronomy is discovered.  Just extra legwork & cpu cylces on the mod side.

Anyway thanks for looking into it.

**Comment by Tekamthi** [comment]
I've just added the following promotion to my mod:

`INSERT INTO UnitPromotions
		(Type,							Description,							Help,										PortraitIndex,	IconAtlas,			Sound,				CannotBeChosen,	LostWithUpgrade,	PediaType,		PediaEntry,									ShowInUnitPanel)
VALUES	('PROMOTION_NO_EARLY_OCEAN',	'TXT_KEY_PROMOTION_NO_EARLY_OCEAN',		'TXT_KEY_PROMOTION_NO_EARLY_OCEAN_HELP',	50,				'extraPromo_Atlas',	'AS2D_IF_LEVELUP',	1,				1,					'PEDIA_SHARED',	'TXT_KEY_PEDIA_PROMOTION_NO_EARLY_OCEAN',	1);

INSERT INTO Language_en_US 
		(Tag,										Text)
VALUES	('TXT_KEY_PROMOTION_NO_EARLY_OCEAN',		'[COLOR_NEGATIVE_TEXT]Ocean Impassable before Astronomy[ENDCOLOR]'),
		('TXT_KEY_PROMOTION_NO_EARLY_OCEAN_HELP',	'Unit cannot enter ocean until the discovery of astronomy'),
		('TXT_KEY_PEDIA_PROMOTION_NO_EARLY_OCEAN',	'Ocean Impassable before Astronomy');

INSERT INTO UnitPromotions_Terrains
		(PromotionType,					TerrainType,		Impassable)		
VALUES	('PROMOTION_NO_EARLY_OCEAN',	'TERRAIN_OCEAN',	true);

INSERT INTO UnitPromotions_Features
        (PromotionType,					FeatureType,						Impassable)
VALUES	('PROMOTION_NO_EARLY_OCEAN',	'FEATURE_RESTRICTED_VISIBILITY',	true);`

I have it set to "Impassable" for both the ocean terrain, as well as the specific feature at issue & assigned this to all embarked units.... no effect!  pre-astro embarked unit can still access ocean with this feature :(

**Comment by Tekamthi** [comment]
aye, the modmod feature that caught my attention was only implemented a few months ago, thus I only noticed recently and can't say if this was always a problem or the result of some change.  While I can make sense of simple things in the .dll, tracking down all the relevant code involved here remains beyond my capacities.

For now I have "solved" this in my modmod (yet unpublished) via a lua function to check all embarked units at end of turn, and move any unit in ocean that shouldn't be back towards capital until its in coast again.  My function also deletes the exploited feature (this fits cuz my storms are meant to be temporary anyway, but wouldn't in the case of an ocean atoll).  Imperfect fix in any case, and sightly wasteful of resources, but this at least stops players from discovering/attacking far off lands they shouldn't be able to.

**Comment by azum4roll** [comment]
Ocean is just special.
![image](https://github.com/LoneGazebo/Community-Patch-DLL/assets/14953746/0cf2f642-0b09-45c4-a761-f31f1bfa95e2)

**Comment by Tekamthi** [comment]
just a reminder from above, maybe you guys are on it already -- but its also mountains that have this weird effect, not just oceans: add a feature of some kind to mountain and suddenly units without mountain access can traverse

**Comment by Tekamthi** [comment]
small update here -- i've vanilla-fied my storms mod, and am now testing it there -- in vanilla the opposite problem exists -- embarked units cannot enter storms but naval units can!
